### PR TITLE
refactor to avoid scope.

### PR DESCRIPTION
**Pull request by panyx0718**
test=develop

### CODE REVIEW

File: paddle/fluid/imperative/layer.cc [hunk 1]
```diff
@@ -85,26 +85,25 @@ class Autograd {
...
-
-  framework::Scope* scope_;
 };
 
-framework::Variable* CreateVariable(const std::string& name,
-                                    const framework::DDim& dim, float val,
-                                    framework::Scope* scope,
-                                    bool random_name = true) {
+void CreateVariable(const std::string& name, const framework::DDim& dim,
+                    float val, bool random_name, framework::Variable* var) {
```
Comment (velconia) [inline paddle/fluid/imperative/layer.cc]: ``` random_name ``` should be removed

File: paddle/fluid/imperative/layer.cc [hunk 2]
```diff
@@ -116,107 +115,74 @@ framework::Variable* CreateVariable(const std::string& name,
...
-  grads_ = CreateVariable(framework::GradVarName(var_desc_->Name()),
-                          var_->Get<framework::LoDTensor>().dims(), 1.0, scope,
-                          false);
+void VarBase::RunBackward() {
   if (!pre_op_) return;
-  Autograd(scope).RunBackward(this);
+
+  auto grads_t = grads_->GetMutable<framework::LoDTensor>();
+  float* data = grads_t->mutable_data<float>(platform::CPUPlace());
+  std::fill(data, data + grads_t->numel(), 1.0);
```
Comment (velconia) [inline paddle/fluid/imperative/layer.cc:160]: Consider the GPU part, should we append a ```fill_constant``` op here?

File: paddle/fluid/imperative/layer.h [hunk 3]
```diff
@@ -33,18 +86,26 @@ class VarBase {
...
-        var_(nullptr),
-        grads_(nullptr) {}
-
-  virtual ~VarBase() {}
-
-  void ApplyGrad(framework::Scope* scope, framework::Variable* grad);
+        var_(new framework::Variable()),
+        grads_(new framework::Variable()) {}
+
+  virtual ~VarBase() {
```
Comment (velconia) [inline paddle/fluid/imperative/layer.h:91]: Better change from raw pointer to std::unique_ptr, then delete operation could be avoided

File: paddle/fluid/operators/fill_constant_op.cc [hunk 4]
```diff
@@ -12,68 +12,40 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-#include "paddle/fluid/framework/data_type.h"
-#include "paddle/fluid/framework/op_registry.h"
-#include "paddle/fluid/operators/math/math_function.h"
+#include "paddle/fluid/operators/fill_constant_op.h"
 
 namespace paddle {
 namespace operators {
 
-class FillConstantInferShape : public framework::InferShapeBase {
+class FillConstantOp : public framework::OperatorWithKernel {
```
Comment (velconia) [inline paddle/fluid/operators/fill_constant_op.cc:20]: Why this change related to autograd?

File: paddle/fluid/imperative/layer.cc [hunk 5]
```diff
@@ -116,107 +115,74 @@ framework::Variable* CreateVariable(const std::string& name,
...
+  PreparedOp p = PreparedOp::Prepare(ctx, *op_kernel, place);
+  p.op.RuntimeInferShape(scope, place, ctx);
+  p.func(framework::ExecutionContext(p.op, scope, *p.dev_ctx, p.ctx));
+
+  for (auto it : grad_output_vars_) {
+    auto& outputs = grad_outputs[it.first];
+    auto& origin_outputs = it.second;
+    for (size_t i = 0; i < outputs.size(); ++i) {
+      framework::Variable* orig_grad = origin_outputs[i];
+      AddTo(outputs[i], orig_grad);
```
Comment (velconia) [inline paddle/fluid/imperative/layer.cc:149]: Why not ```TensorCopy``` here?

File: paddle/fluid/imperative/layer.h [hunk 6]
```diff
@@ -14,17 +14,70 @@
...
+    if (kernels_iter == all_op_kernels.end()) {
+      PADDLE_THROW(
+          "There are no kernels which are registered in the %s operator.",
+          op.Type());
+    }
+
+    framework::OperatorWithKernel::OpKernelMap& kernels = kernels_iter->second;
+
+    auto expected_kernel_key = op.GetExpectedKernelType(
+        framework::ExecutionContext(op, dummy_scope, *dev_ctx, ctx));
```
Comment (velconia) [inline paddle/fluid/imperative/layer.h]: framework::Scope()

File: paddle/fluid/imperative/layer.cc [hunk 7]
```diff
@@ -116,107 +115,74 @@ framework::Variable* CreateVariable(const std::string& name,
...
-    }
-    if (!found) {
-      ret.push_back(nullptr);
+  framework::OperatorWithKernel* op_kernel =
+      dynamic_cast<framework::OperatorWithKernel*>(opbase.get());
+  PADDLE_ENFORCE_NOT_NULL(op_kernel, "only support op with kernel");
+
+  framework::Scope scope;
+  platform::CPUPlace place;
+  PreparedOp p = PreparedOp::Prepare(ctx, *op_kernel, place);
```
Comment (velconia) [inline paddle/fluid/imperative/layer.cc:140]: This part LGTM

File: paddle/fluid/imperative/layer.cc (hunk 2 above)
Comment (panyx0718) [reply paddle/fluid/imperative/layer.cc:160]: We can leave this as future work. Currently, we only consider CPU

File: paddle/fluid/imperative/layer.cc (hunk 1 above)
Comment (panyx0718) [reply paddle/fluid/imperative/layer.cc]: Done

File: paddle/fluid/imperative/layer.h (hunk 3 above)
Comment (panyx0718) [reply paddle/fluid/imperative/layer.h:91]: unique_ptr is not pybind friendly.

File: paddle/fluid/operators/fill_constant_op.cc (hunk 4 above)
Comment (panyx0718) [reply paddle/fluid/operators/fill_constant_op.cc:20]: We only call OperatorWithKernel. FillConstant was OperatorBase, changed to OperatorWithKernel

File: paddle/fluid/imperative/layer.cc (hunk 5 above)
Comment (panyx0718) [reply paddle/fluid/imperative/layer.cc:149]: this is `a += b`, not `a = b`

File: paddle/fluid/imperative/layer.h (hunk 6 above)
Comment (panyx0718) [reply paddle/fluid/imperative/layer.h]: Done

**Review by velconia — APPROVED** [review]
LGTM
